README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# whatr <img src='man/figures/logo.png' align="right" height="139" />

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://lifecycle.r-lib.org/articles/stages.html#maturing)
[![CRAN
status](https://www.r-pkg.org/badges/version/whatr)](https://CRAN.R-project.org/package=whatr)
![Downloads](https://cranlogs.r-pkg.org/badges/grand-total/whatr)
[![Codecov test
coverage](https://img.shields.io/codecov/c/github/kiernann/whatr/master.svg)](https://codecov.io/gh/kiernann/whatr?branch=master')
[![R build
status](https://github.com/kiernann/whatr/workflows/R-CMD-check/badge.svg)](https://github.com/kiernann/whatr/actions)
<!-- badges: end -->

> *This* R package was made to facilitate the analysis of game show data
> by scraping the [J! Archive](https://j-archive.com/).

> What is… whatr?

## Installation

The release version of ‘whatr’ can be installed from
[CRAN](https://cran.r-project.org/package=whatr):

``` r
install.packages("whatr")
```

You can install the development version of ‘whatr’ from
[GitHub](https://github.com/kiernann/whatr) with:

``` r
# install.packages("remotes")
remotes::install_github("kiernann/whatr")
```

## Usage

The `whatr_*()` functions take one of four inputs to find the
appropriate game on the J! Archive:

1.  An HTML document object from `whatr_html()` or similar.
2.  The numeric, non-sequential game ID.
3.  The sequential show number, as character starting with `#`.
4.  The date an episode originally aired, as `yyyy-mm-dd`.

``` r
library(whatr)
page <- whatr_html("2019-06-03")
whatr_board(page)[, 5:7]
#> # A tibble: 61 x 3
#>    category       clue                                                         answer              
#>    <chr>          <chr>                                                        <chr>               
#>  1 Picture The I… For An Optimistic View Of Things, Take A Look Through These  Rose-Colored Glasses
#>  2 Picture The I… If You Have These, It Means You're Well-Organized            Your Ducks In A Row 
#>  3 Picture The I… It's What's Going On Here                                    Comparing Apples & …
#>  4 Picture The I… Greenhouse Gas Emission Due To Human Activity Leaves Behind… A Carbon Footprint  
#>  5 Picture The I… Note The Lovely Weather; It Means To Take Advantage Of An O… Make Hay While The …
#>  6 Literature     Melville's 'Bartleby The Scrivener' Is Subtitled 'A Story O… Wall Street         
#>  7 Literature     In A 1923 Book By Kahlil Gibran, Almustafa Is This Mystical… The Prophet         
#>  8 Literature     In 'Charlotte's Web', Templeton Is This Creature             A Rat               
#>  9 Literature     In A Novel Simone De Beauvoir Depicted Herself As Anne & Th… (Albert) Camus      
#> 10 Literature     The Title Peak Of This Thomas Mann Novel Is Home To A Swiss… Magic Mountain      
#> # … with 51 more rows
whatr_scores(6304)
#> # A tibble: 63 x 5
#>    round     i name  score double
#>    <int> <int> <chr> <int> <lgl> 
#>  1     1     1 James  1000 TRUE  
#>  2     1     2 Emma   1000 FALSE 
#>  3     1     3 James   800 FALSE 
#>  4     1     4 James  1000 FALSE 
#>  5     1     5 Emma   1000 FALSE 
#>  6     1     6 Jay     800 FALSE 
#>  7     1     7 James   600 FALSE 
#>  8     1     8 James  1000 FALSE 
#>  9     1     9 James  1000 FALSE 
#> 10     1    10 Emma    800 FALSE 
#> # … with 53 more rows
whatr_plot("#8006")
```

<img src="man/figures/README-usage-1.png" width="100%" />

## Data

Included in this package is data on the 217 episodes of the show’s 35th
season.

``` r
whatr::episodes
#> # A tibble: 217 x 3
#>     game  show date      
#>    <int> <int> <date>    
#>  1  6096  7816 2018-09-10
#>  2  6097  7817 2018-09-11
#>  3  6098  7818 2018-09-12
#>  4  6099  7819 2018-09-13
#>  5  6101  7820 2018-09-14
#>  6  6102  7821 2018-09-17
#>  7  6103  7822 2018-09-18
#>  8  6105  7823 2018-09-19
#>  9  6106  7824 2018-09-20
#> 10  6107  7825 2018-09-21
#> # … with 207 more rows
whatr::synopses
#> # A tibble: 651 x 6
#>     game name    final coryat right wrong
#>    <int> <chr>   <int>  <int> <int> <int>
#>  1  6096 Rick    20200  18000    23     2
#>  2  6096 Isaac    5200   6400    11     4
#>  3  6096 Caitlin  8800   7800    13     3
#>  4  6097 Rick    29199  17400    24     2
#>  5  6097 Lori    34401  14800    20     2
#>  6  6097 Becky    3200   3200     9     3
#>  7  6098 Lori    17201  14200    18     0
#>  8  6098 Kyle    28401  17200    21     3
#>  9  6098 Monica   8400   6000    14     5
#> 10  6099 Kyle     9599  16000    20     3
#> # … with 641 more rows
```

------------------------------------------------------------------------

The ‘whatr’ project is released with a [Contributor Code of
Conduct](https://kiernann.com/whatr/CODE_OF_CONDUCT.html). By
contributing to this project, you agree to its terms.

The J! Archive is created by fans, for fans. The *Jeopardy!* game show
and all elements thereof, including but not limited to copyright and
trademark thereto, are the property of Jeopardy Productions, Inc. and
are protected under law. This package is not affiliated with, sponsored
by, or operated by Jeopardy Productions, Inc or the J! Archive itself.

<!-- refs: start -->
<!-- refs: end -->


NEWS.md
# whatr 1.1.0

# whatr 1.0.1

* Improve reliability of tests to pass at any time.
* Improve how tests randomly wait before running.
* Remove non-ASCII characters from data.

# whatr 1.0.0

* Fix player scraping for season 35 multi-team episodes (#7).
* Rename `whatr_info()` to `whatr_airdate()`.
* Remove `whatr_id()` in favor of `whatr_html()`.
* Rename the `n` column to `i` to avoid confusion with `count()`, etc.
* Stop splitting the player `from` column into `city` and `state`.
* Finish documentaing all the functions.
* Use spelling.

# whatr 0.5.0

* Rename `whatr_info()` to `whatr_airdate()`.
* Remove `whatr_id()` in favor of `whatr_html()`.
* Rename the `n` column to `i` to avoid confusion with `count()`, etc.
* Stop splitting the player `from` column into `city` and `state`.
* Finish documenting all the functions.

# whatr 0.4.1

* Use new `entity_clean()` for all `html_text()`.
* Use `html_attr()` for final answer too.

# whatr 0.4.0

* Rename `whatr_summary()` to `whatr_synopsis()`.
* Add data from the 35th season of the show.

# whatr 0.3.0

* Use _new_, new `whatr_html()` method as input for all functions.
* Add `whatr_doubles()`

# whatr 0.2.0

* Use new `whatr_html()` method as optional input for all functions.
* Add `whatr_plot()`.

# whatr 0.1.0

* Add all functions with proper pipe formatting.
  * `whatr_data()`
  * `whatr_info()`
  * `whatr_summary()`
  * `whatr_players()`
  * `whatr_scores()`
  * `whatr_board()`
      * `whatr_categories()`
      * `whatr_clues()`
      * `whatr_answers()`

# whatr 0.0.2

* Added basic functionality for scores.
* Added a `NEWS.md` file to track changes to the package.
* Removed all other functions, need to code as `dplyr` pipes.


R/whatr-html.R
#' Return the right J! Archive HTML document
#'
#' To avoid downloading and reading the J! Archive over and over, this function
#' can be used to return an HTML document from one of four types of input:
#' 1. Game ID (6304)
#' 2. Show number ("#8006")
#' 3. Air date ("2019-06-03")
#' 4. HTML document itself
#'
#' @param x Any kind of J! Archive game identifier.
#' @param out One of "showscores" or "showgame" used for [whatr_scores()] or
#'   [whatr_clues()], etc. respectively. Either document can be returned from
#'   the input of the other.
#' @return A J! Archive `html_document`.
#' @examples
#' whatr_html(x = 6304, out = "showscores")
#' whatr_html(x = "2019-06-03", out = "showgame")
#' whatr_html("#8006", "showgame") %>% whatr_clues()
#' @importFrom httr GET content
#' @importFrom methods is
#' @importFrom stringr str_extract str_detect
#' @export
whatr_html <- function(x, out = c("showgame", "showscores")) {
  out <- match.arg(out, c("showgame", "showscores"))
  if (is.numeric(x)) {
    html <- httr::GET(
      url = sprintf("https://www.j-archive.com/%s.php", out),
      query = list(game_id = x)
    )
    # message(sprintf("in:  id %s\nout: %s", x, out))
    httr::content(html)
  } else if (is(x, "xml_document")) {
    c <- as.character(x)
    if (stringr::str_detect(c, "ddred") & out == "showscores") {
      # message("in:  showscores\nout: showscores")
      return(x)
    } else if (stringr::str_detect(c, "ddred") & out == "showgame") {
      # message("in:  showscores\nout: showgame")
      showscores_to_game(html = x)
    } else if (!stringr::str_detect(c, "ddred") & out == "showgame") {
      # message("in:  showgame\nout: showgame")
      return(x)
    } else if (!stringr::str_detect(c, "ddred") & out == "showscores") {
      # message("in:  showgame\nout: showscores")
      showgame_to_scores(html = x)
    }
  } else if (stringr::str_detect(x, "^#\\d+$")) {
    if (out == "showgame") {
      # message(sprintf("in:  show %s\nout: showgame", x))
      show_to_game(show = x)
    } else if (out == "showscores") {
      # message(sprintf("in:  show %s\nout: showscores", x))
      show_to_scores(show = x)
    }
  } else if (stringr::str_detect(x, "^\\d{4}-\\d+-\\d+$")) {
    if (out == "showgame") {
      # message(sprintf("in:  %s\nout: showgame", x))
      date_to_game(date = x)
    } else if (out == "showscores") {
      # message(sprintf("in:  %s\nout: showscores", x))
      date_to_scores(date = x)
    }
  } else {
    stop("not able to determine input type, see ?whatr_html")
  }
}

showgame_to_scores <- function(html) {
  id <- html %>%
    rvest::html_node(".game_dynamics") %>%
    rvest::html_attr("src") %>%
    stringr::str_extract("\\d+$")
  data <- httr::GET(
    url = "https://www.j-archive.com/showscores.php",
    query = list(game_id = id)
  )
  return(httr::content(data))
}

showscores_to_game <- function(html) {
  id <- html %>%
    rvest::html_node("#game_title") %>%
    rvest::html_node("a") %>%
    rvest::html_attr("href") %>%
    stringr::str_extract("\\d+$")
  data <- httr::GET(
    url = "https://www.j-archive.com/showgame.php",
    query = list(game_id = id)
  )
  httr::content(data)
}

date_to_game <- function(date) {
  data <- httr::GET(
    url = "https://www.j-archive.com/search.php",
    query = list(search = paste("date", date, sep = ":"))
  )
  httr::content(data)
}

date_to_scores <- function(date) {
  redirect <- httr::HEAD(
    url = "https://www.j-archive.com/search.php",
    query = list(search = paste("date", date, sep = ":"))
  )
  id <- stringr::str_extract(redirect$url, "\\d+$")
  data <- httr::GET(
    url = "https://www.j-archive.com/showscores.php",
    query = list(game_id = id)
  )
  httr::content(data)
}

show_to_game <- function(show) {
  data <- httr::GET(
    url = "https://www.j-archive.com/search.php",
    query = list(search = paste("show", gsub("#", "", show), sep = ":"))
  )
  httr::content(data)
}

show_to_scores <- function(show) {
  redirect <- httr::HEAD(
    url = "https://www.j-archive.com/search.php",
    query = list(search = paste("show", gsub("#", "", show), sep = ":"))
  )
  id <- stringr::str_extract(redirect$url, "\\d+$")
  data <- httr::GET(
    url = "https://www.j-archive.com/showscores.php",
    query = list(game_id = id)
  )
  httr::content(data)
}
